File: EmergencyMutualAidInformation/backend/controllers/user_controller.go
package controllers

import (
	"emergency-mutual-aid/database"
	"emergency-mutual-aid/models"
	"emergency-mutual-aid/utils"
	"strconv"

	"github.com/gin-gonic/gin"
	"golang.org/x/crypto/bcrypt"
)

func UpdateProfile(c *gin.Context) {
	userID := c.GetUint("user_id")

	var req struct {
		Email    string `json:"email"`
		Phone    string `json:"phone"`
		RealName string `json:"real_name"`
		Avatar   string `json:"avatar"`
	}

	if err := c.ShouldBindJSON(&req); err != nil {
		utils.BadRequest(c, "参数错误")
		return
	}

	var user models.User
	if err := database.DB.First(&user, userID).Error; err != nil {
		utils.NotFound(c, "用户不存在")
		return
	}

	if req.Email != "" {
		user.Email = req.Email
	}
	if req.Phone != "" {
		user.Phone = req.Phone
	}
	if req.RealName != "" {
		user.RealName = req.RealName
	}
	if req.Avatar != "" {
		user.Avatar = req.Avatar
	}

	if err := database.DB.Save(&user).Error; err != nil {
		utils.InternalServerError(c, "更新失败")
		return
	}

	utils.Success(c, nil)
}

func GetFavorites(c *gin.Context) {
	userID := c.GetUint("user_id")

	var favorites []models.Favorite
	if err := database.DB.Where("user_id = ?", userID).Find(&favorites).Error; err != nil {
		utils.InternalServerError(c, "获取收藏列表失败")
		return
	}

	utils.Success(c, favorites)
}

func AddFavorite(c *gin.Context) {
	userID := c.GetUint("user_id")

	var req struct {
		Type     string `json:"type" binding:"required"`
		TargetID uint   `json:"target_id" binding:"required"`
	}

	if err := c.ShouldBindJSON(&req); err != nil {
		utils.BadRequest(c, "参数错误")
		return
	}

	var existing models.Favorite
	if err := database.DB.Where("user_id = ? AND type = ? AND target_id = ?", userID, req.Type, req.TargetID).First(&existing).Error; err == nil {
		utils.BadRequest(c, "已收藏")
		return
	}

	favorite := models.Favorite{
		UserID:   userID,
		Type:     req.Type,
		TargetID: req.TargetID,
	}

	if err := database.DB.Create(&favorite).Error; err != nil {
		utils.InternalServerError(c, "收藏失败")
		return
	}

	utils.Success(c, nil)
}

func RemoveFavorite(c *gin.Context) {
	userID := c.GetUint("user_id")
	id := c.Param("id")

	if err := database.DB.Where("id = ? AND user_id = ?", id, userID).Delete(&models.Favorite{}).Error; err != nil {
		utils.InternalServerError(c, "取消收藏失败")
		return
	}

	utils.Success(c, nil)
}

func GetUsers(c *gin.Context) {
	page, _ := strconv.Atoi(c.DefaultQuery("page", "1"))
	pageSize, _ := strconv.Atoi(c.DefaultQuery("page_size", "10"))
	keyword := c.Query("keyword")

	offset := (page - 1) * pageSize

	var users []models.User
	query := database.DB.Model(&models.User{})

	if keyword != "" {
		query = query.Where("username LIKE ? OR email LIKE ? OR real_name LIKE ?", "%"+keyword+"%", "%"+keyword+"%", "%"+keyword+"%")
	}

	var total int64
	query.Count(&total)

	if err := query.Offset(offset).Limit(pageSize).Order("created_at DESC").Find(&users).Error; err != nil {
		utils.InternalServerError(c, "获取用户列表失败")
		return
	}

	utils.Success(c, gin.H{
		"total": total,
		"page":  page,
		"list":  users,
	})
}

func GetUser(c *gin.Context) {
	id := c.Param("id")

	var user models.User
	if err := database.DB.First(&user, id).Error; err != nil {
		utils.NotFound(c, "用户不存在")
		return
	}

	utils.Success(c, user)
}

func CreateUser(c *gin.Context) {
	var user models.User
	if err := c.ShouldBindJSON(&user); err != nil {
		utils.BadRequest(c, "参数错误")
		return
	}

	var existing models.User
	if err := database.DB.Where("username = ?", user.Username).First(&existing).Error; err == nil {
		utils.BadRequest(c, "用户名已存在")
		return
	}

	if user.Password == "" {
		utils.BadRequest(c, "密码不能为空")
		return
	}

	hashedPassword, err := bcrypt.GenerateFromPassword([]byte(user.Password), bcrypt.DefaultCost)
	if err != nil {
		utils.InternalServerError(c, "密码加密失败")
		return
	}
	user.Password = string(hashedPassword)
	user.Status = 1
	if user.Role == "" {
		user.Role = "user"
	}

	if err := database.DB.Create(&user).Error; err != nil {
		utils.InternalServerError(c, "创建用户失败")
		return
	}

	user.Password = ""
	utils.Success(c, user)
}

func UpdateUser(c *gin.Context) {
	id := c.Param("id")

	var user models.User
	if err := database.DB.First(&user, id).Error; err != nil {
		utils.NotFound(c, "用户不存在")
		return
	}

	var req struct {
		Email    string `json:"email"`
		Phone    string `json:"phone"`
		RealName string `json:"real_name"`
		Avatar   string `json:"avatar"`
		Status   int    `json:"status"`
		Role     string `json:"role"`
	}

	if err := c.ShouldBindJSON(&req); err != nil {
		utils.BadRequest(c, "参数错误")
		return
	}

	if req.Email != "" {
		user.Email = req.Email
	}
	if req.Phone != "" {
		user.Phone = req.Phone
	}
	if req.RealName != "" {
		user.RealName = req.RealName
	}
	if req.Avatar != "" {
		user.Avatar = req.Avatar
	}
	if req.Status != 0 {
		user.Status = req.Status
	}
	if req.Role != "" {
		user.Role = req.Role
	}

	if err := database.DB.Save(&user).Error; err != nil {
		utils.InternalServerError(c, "更新用户失败")
		return
	}

	utils.Success(c, user)
}

func DeleteUser(c *gin.Context) {
	id := c.Param("id")

	if err := database.DB.Delete(&models.User{}, id).Error; err != nil {
		utils.InternalServerError(c, "删除用户失败")
		return
	}

	utils.Success(c, nil)
}

File: PhotographyStudioAppointment/backend/src/controllers/userController.js
const { Op } = require('sequelize');
const { User } = require('../models');

exports.list = async (req, res) => {
  try {
    const { page = 1, pageSize = 10, keyword = '', role = '' } = req.query;
    const offset = (page - 1) * pageSize;

    const where = {};
    if (keyword) {
      where[Op.or] = [
        { username: { [Op.like]: `%${keyword}%` } },
        { name: { [Op.like]: `%${keyword}%` } },
        { phone: { [Op.like]: `%${keyword}%` } }
      ];
    }
    if (role) {
      where.role = role;
    }

    const { count, rows } = await User.findAndCountAll({
      where,
      attributes: { exclude: ['password'] },
      offset: parseInt(offset),
      limit: parseInt(pageSize),
      order: [['createdAt', 'DESC']]
    });

    res.json({
      total: count,
      list: rows,
      page: parseInt(page),
      pageSize: parseInt(pageSize)
    });
  } catch (error) {
    console.error('List users error:', error);
    res.status(500).json({ message: '服务器内部错误' });
  }
};

exports.getStaff = async (req, res) => {
  try {
    const { role } = req.query;
    const where = { status: 'active' };
    if (role) {
      where.role = role;
    }

    const users = await User.findAll({
      where,
      attributes: ['id', 'name', 'role', 'phone'],
      order: [['name', 'ASC']]
    });

    res.json(users);
  } catch (error) {
    console.error('Get staff error:', error);
    res.status(500).json({ message: '服务器内部错误' });
  }
};

exports.create = async (req, res) => {
  try {
    const { username, password, name, phone, email, role, status } = req.body;

    const existing = await User.findOne({ where: { username } });
    if (existing) {
      return res.status(400).json({ message: '用户名已存在' });
    }

    const user = await User.create({
      username,
      password,
      name,
      phone,
      email,
      role,
      status
    });

    const { password: _, ...userInfo } = user.toJSON();
    res.status(201).json(userInfo);
  } catch (error) {
    console.error('Create user error:', error);
    res.status(500).json({ message: '服务器内部错误' });
  }
};

exports.update = async (req, res) => {
  try {
    const { id } = req.params;
    const { password, ...updateData } = req.body;

    const user = await User.findByPk(id);
    if (!user) {
      return res.status(404).json({ message: '用户不存在' });
    }

    if (password) {
      user.password = password;
    }
    Object.assign(user, updateData);
    await user.save();

    const { password: _, ...userInfo } = user.toJSON();
    res.json(userInfo);
  } catch (error) {
    console.error('Update user error:', error);
    res.status(500).json({ message: '服务器内部错误' });
  }
};

exports.remove = async (req, res) => {
  try {
    const { id } = req.params;

    if (parseInt(id) === req.user.id) {
      return res.status(400).json({ message: '不能删除当前登录用户' });
    }

    const user = await User.findByPk(id);
    if (!user) {
      return res.status(404).json({ message: '用户不存在' });
    }

    await user.destroy();
    res.json({ message: '删除成功' });
  } catch (error) {
    console.error('Delete user error:', error);
    res.status(500).json({ message: '服务器内部错误' });
  }
};
